### PR TITLE
Function extension mechanism and several functions

### PR DESCRIPTION
**Pull request by cabo**

### CODE REVIEW

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1212]: s /name-first/function-name-first/

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1216]: Do we want to allow more whitespace, e.g. just inside the parentheses?

Comment (glyn) [inline draft-ietf-jsonpath-base.md:1216]: The following is more consistent with the rest of our ABNF:
```suggestion
function-expression     = function-name "(" [function-argument *("," S function-argument)] ")"
```

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1217]: Aside: I like incremental alternatives and if we keep them here, we may want to use them elsewhere, e.g. in the definition of `selector`.

Comment (glyn) [inline draft-ietf-jsonpath-base.md:1217]: If a function returns a nodelist, how can the result be used as a singular path?

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1225]: If we fix #195, I think the language here can be simplified. Until then, we need to include `exist-expr` in the places where a singular node must be returned.

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1248]: It seems we have limited options if we are going to be consistent with the rest of the spec. We could either return a fixed value (e.g. `-1`) or an empty nodelist. The latter seems preferable as it's consistent with a singular path possibly being empty.

This makes me wonder whether we should make a function always return a nodelist and have the function spec say whether or not the returned nodelist is "singular" (meaning it has at most one member) instead of allowing nodelists of values to be returned.

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1673]:
```suggestion
function extension names, keeping them in reserve for applications that
```

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1224]: Rather than a singular node, perhaps return a singular nodelist (see below).

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1207]: Perhaps rather than the item being  a value, it could be a singular nodelist. This allows chaining the result of a function as an argument of another function with propagation of the empty nodelist. Whether we make a general rule that empty argument nodelists MUST produce empty result nodelists or whether this is a per-function optional behaviour is something we should decide. I quite like the uniformity and predictability of a general rule.

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1212]:
```suggestion
function-name-char      = DIGIT / function-name-first / "_"
```

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1216]: space inside parentheses is evil.  Except for line breaks...

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1217]: This was an expedient device to fit this in without touching everything -- should be fixed IMHO

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1248]: The main problem is that the node that has the length isn't in the input JSON value

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1674]: Not sure about "can make good use of their shortness". When would a short name not be made good use of?

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1678]: This may be common policy for IANA, but the risk is that a spec may never appear and we end up with people "squatting" on function names so others can't register them.

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1672]: Is there just one expert? Why not:
```suggestion
Experts are instructed to be frugal in the allocation of very short
```

Comment (cabo) [reply draft-ietf-jsonpath-base.md:1672]:
```suggestion
The experts is instructed to be frugal in the allocation of very short
```

I'd say.

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1712]: Do we need to allow an empty nodelist as input so that we can apply `length` to a `singular-path`?

**Review by glyn — COMMENTED** [review]
An excellent start, thanks! Comments provided.

File: draft-ietf-jsonpath-base.md
Comment (glyn) [reply draft-ietf-jsonpath-base.md:1248]: Of course. I was thinking "list of nodes" rather than _nodelist_.

Comment (glyn) [reply draft-ietf-jsonpath-base.md:1248]: If we can avoid the function having to create nodes, so much the better. So I guess that limits return values to nodelists or values after all.

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [inline draft-ietf-jsonpath-base.md:1233]: I don't know that this is need in the spec.  Might be good for a comment, but its target audience is really people who write specs.

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [inline draft-ietf-jsonpath-base.md:1245]: Do we also want to support strings or objects?  They're fairly well-defined for this function.

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [inline draft-ietf-jsonpath-base.md:1673]: What do you mean by "applications" here?

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [reply draft-ietf-jsonpath-base.md:1678]: Perhaps the function name reservation can have an expiry.

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [inline draft-ietf-jsonpath-base.md:1690]: So we're only accepting English-character function names?  Just clarifying.

**Review by gregsdennis — CHANGES_REQUESTED** [review]

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1233]: (Maybe I should have added a "to me removed by RFC editor" comment.)

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1245]: Very good question.  What do implementations out there do?
For strings, I think we'd count the Unicode scalars.
For JSON objects, the number of name-value pairs.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1673]: application in the sense of the way this is going to the used, usage

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1690]: Well, ASCII, yes.
This is an interesting question in a more general way, as so far JSONPath is mostly indifferent to languages.
But these function extensions are intended for an international audience, so it is probably best to do a lowest common denominator.

**Comment by cabo** [comment]
Weirdly, I cannot reply to this observation: https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/280#discussion_r1008936675

Expiry is somewhat unusual for IANA registries (early allocations do expire).
What is the behavior you would expect implementations to exhibit?
The function extension to stop working on the expiry day?

File: draft-ietf-jsonpath-base.md
Comment (goessner) [reply draft-ietf-jsonpath-base.md:1207]: Shouldn't the function definition being able to define the "return" even more concretely as a string or number? The latter would be, what is expected with our `length` function.

File: draft-ietf-jsonpath-base.md
Comment (goessner) [reply draft-ietf-jsonpath-base.md:1245]: We even could allow nodelists and return the number of nodes.

File: draft-ietf-jsonpath-base.md
Comment (goessner) [reply draft-ietf-jsonpath-base.md:1248]: I like the well known return value `-1` for the case of an invalid argument.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1245]: That would need to be a separate extension, as length(X) is ambiguous if X can be interpreted as a node that contains an array (which gets us the size of the array) or a nodelist of size 1 (which gets us 1).

File: draft-ietf-jsonpath-base.md
Comment (glyn) [reply draft-ietf-jsonpath-base.md:1207]: Also, what happens when a function is passed an argument of the wrong kind? Does this produce a syntax error? Is there any coercion to the correct kind? Are there any "soft" runtime errors in which case what does the function return?

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [reply draft-ietf-jsonpath-base.md:1673]: I wonder if another word would work better here so that readers don't conflate it with executable applications or other consuming code.

Comment (glyn) [reply draft-ietf-jsonpath-base.md:1673]: Perhaps:
```suggestion
function extension names, keeping them in reserve for functions that
```

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1690]: Obviously, I stole the ABNF from dot-member-name, which allows %x80-10FFFF -- this is a recipe for confusion, with RTL characters, format effectors, and all that.  I'm having a hard time making a suggestion as to what we should allow there -- `\p{L}|\p{N}` maybe, in parallel to allowing ALPHA + DIGIT?  But then we also allow "_", and there may be equivalent separators in some scripts.  So this is hard to get useful over a wide domain, while reducing the potential for confusion.  function names of course do not depend on whatever language was used in the input data item, so we can be more frugal in what we allow.

**Comment by timbray** [comment]
Hey @cabo, based on this useful-looking discussion, do you plan to revise this significantly?  I'm wondering if I should give it a careful read now or wait for an updated version.

**Comment by cabo** [comment]
Definitely.  Of course, I'd like to have your input on the points raised (the non-trivial ones).  We might even want to have a design-team meeting...

**Comment by timbray** [comment]
> Definitely. Of course, I'd like to have your input on the points raised (the non-trivial ones). We might even want to have a design-team meeting...

Might be a good idea to mark the non-non-trivial issues above resolved, to help reviewer focus. If it's OK, I'll wait for the next revision to go end-to-end. Agreed on a design meeting.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1674]: Indeed, this is really less about shortness.  Say, someone registers a size() that extracts the shirt size of a JSON description of a shirt.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1678]: Finally, I can reply to this observation.
Copying from below:

Expiry is somewhat unusual for IANA registries (early allocations do expire).
What is the behavior you would expect implementations to exhibit?
The function extension to stop working on the expiry day?

**Comment by cabo** [comment]
Here's my list of items where we cold use input:

- More frugal with allowing whitespace?
- return value of functions that don't get the right arguments (e.g., length(1))
- (interaction with #195 -- non-singular paths in exist-expr)
- IANA designated expert instructions (frugal with "good" names)
- expiring reservations?  (shades of certificates)
- domain of length (arrays now, add strings, objects?)?
- should there be another function for the size of nodelists (size())?

Probably mostly resolved:
- `[a-z][_a-z0-9]*` for function names (ASCII, lowercase, _ only)

File: draft-ietf-jsonpath-base.md
Comment (goessner) [reply draft-ietf-jsonpath-base.md:1225]: I don't think, that #195 is relevant here, since `exist-expr` always results in a Boolean, regardless of the presence of a singular-path or not.

Comment (glyn) [reply draft-ietf-jsonpath-base.md:1225]: My point was about an `exist-expr` which tests the existence of the results of a function. Unless/until #195 is merged, a function used in such a context MUST return a singular node.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md]: Well, the answer is yes; we should move any attempt to reduce variability in spacing to its own PR.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md]: Removed in f7295bd

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1685]: No further discussion -- I'm assuming we stick with `[a-z][_a-z0-9]*` (no change required)

File: draft-ietf-jsonpath-base.md
Comment (glyn) [reply draft-ietf-jsonpath-base.md:1670]: Actually, why not simply the following?
```suggestion
are likely to enjoy wide use.
```
You see I'm not sure the phrase "and can make good use of their shortness" is anything more than motherhood.

File: draft-ietf-jsonpath-base.md
Comment (glyn) [reply draft-ietf-jsonpath-base.md:1674]: @gregsdennis I think those questions are for you. (I wasn't suggesting expiry.)

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md]: Addressed in 85e5525

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1676]: I think the DE instructions are open enough so that the DE can be quite insisting on a specification, so I don't see such a potential for disaster here.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1710]: That should be a separate function (for a singular node list, it is not clear whether you want 1 or the length of the value in the node).

Comment (cabo) [reply draft-ietf-jsonpath-base.md:1710]: ABNF says:

```
function-argument       = comparable
```

So literals are the other case.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1219]: That is certainly a part that we still need to address.
Functions that return a nodelist probably need to be categorized into singular ones and ones that are not.
Let's to that when we agree on the overall architecture of the extension.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1227]: Let's revisit this with the complete proposal

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1245]: I made a different proposal specifically for length(), but we do need to come up with a more general solution.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1208]: Let's see which mistakes can be caught at compile time, and whether we can stick with "no runtime errors".

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1710]: Separate function now in my proposal update.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1242]: Separate function (called "count") provided now.

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1294]: Perhaps add a remdup/dedup function?

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1269]: Does that include the case of a singular path which returns an empty nodelist?

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1294]: Maybe.  Maybe that is a bit too much invention for the base standard.  I myself don't have an opinion on this.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1269]: That is a very good question.  As defined, the argument value needs to be a value, so handling empty nodelists is not defined.  But it should be possible to do this...

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [inline draft-ietf-jsonpath-base.md:1272]: What's the difference between `count()` and `length()`?  Why have a distinction?

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [inline draft-ietf-jsonpath-base.md:1297]: Since we have `match()` and `search()` now, we should probably remove `=~`.

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [inline draft-ietf-jsonpath-base.md:1345]: Need examples of `match()` and `search()` to better highlight the difference between them.  The text varies only by one saying "contains a match" vs just "matches," which I don't think is clear.  In my experience with regexes, containing a match _is_ matching.

**Review by gregsdennis — CHANGES_REQUESTED** [review]

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1286]: This seems syntactically invalid since any function argument has to be a `comparable` which `@.*.author` isn't.

File: draft-ietf-jsonpath-base.md
Comment (glyn) [reply draft-ietf-jsonpath-base.md:1297]: It has mostly gone. There are just two vestigial mentions left.

Comment (cabo) [reply draft-ietf-jsonpath-base.md:1297]: Fixed in 52ca26b

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1345]: Right.  The one example that has a match() would work the same with a search()... we probably need to set up a pair of contrasting examples.  Let's do a round through the examples once the technical content has stabilized.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1286]: Yes.  This doesn't have the results of merging #195 incorporated.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1272]:
``` 
###### Function Extensions: length {#length}
{: unnumbered}

Arguments:
: 1. value

###### Function Extensions: count {#count}
{: unnumbered}

Arguments:
: 1. nodelist
```

One is for JSON values, one is for nodelists.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1346]: The text is fine though.  87d95d6 adds a few words that might help grasping the difference.

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [reply draft-ietf-jsonpath-base.md:1272]: This isn't apparent to me from the text.  Some additional clarification would be good. (May be provided with the examples I asked for in another comment.)

File: draft-ietf-jsonpath-base.md
Comment (gregsdennis) [inline draft-ietf-jsonpath-base.md:1406]: Can we get one with `search()` as well, please?

**Comment by glyn** [comment]
@cabo: I rebased this and resolved some conflicts.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md]: grammar fixed in 08dd7ad
now add semantic restriction

**Comment by cabo** [comment]
Update:

> Here's my list of items where we could use input:
> 
> * More frugal with allowing whitespace?

Address whitespace in a separate PR, if desired.

> * return value of functions that don't get the right arguments (e.g., length(1))

➔ Semantic issue: What to do with value of wrong kind?

> * (interaction with [Do we really require a singular path for `exist-expr` #195](https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/195) -- non-singular paths in exist-expr)

Fixed in 08dd7ad

> * IANA designated expert instructions (frugal with "good" names)

Fixed in afda44a

> * expiring reservations?  (shades of certificates)

Probably not.

> * domain of length (arrays now, add strings, objects?)?

➔ Semantic issue: What to do with value of wrong kind?

> * should there be another function for the size of nodelists (size())?

Yes, that is called count() now (e3a565a)

> Probably mostly resolved:
> 
> * `[a-z][_a-z0-9]*` for function names (ASCII, lowercase, _ only)

Let's stick with that name space.

File: draft-ietf-jsonpath-base.md
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1253]: Why would it not be acceptable to return an empty nodelist since that's valid for use in a comparable?

Comment (glyn) [reply draft-ietf-jsonpath-base.md:1253]: Separately: what should the behaviour be if a nodelist with multiple nodes is returned in this situation?

**Comment by glyn** [comment]
> > * return value of functions that don't get the right arguments (e.g., length(1))
> 
> ➔ Semantic issue: What to do with value of wrong kind?

We may be able to make this syntactically invalid. See https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/320 for some ideas.

File: draft-ietf-jsonpath-base.md
Comment (cabo) [reply draft-ietf-jsonpath-base.md:1253]: One way to handle this would be to build a type system with values, singular paths (which are implicitly converted into values where a value is required), and filter-paths.
Functions already have to define the types of their arguments and their result, so this could be checked statically.

Comment (glyn) [reply draft-ietf-jsonpath-base.md:1253]: Yes, I like that. The more static checks we can get, the fewer runtime errors the spec will need to cope with.

**Review by gregsdennis — APPROVED** [review]

**Pull request merged**
